lib/datadog/tracing/contrib/roda/instrumentation.rb

```diff
@@ -0,0 +1,84 @@
+# frozen_string_literal: true
+
+require_relative '../../metadata/ext'
+require_relative 'ext'
+require_relative '../analytics'
+
+module Datadog
+  module Tracing
+    module Contrib
+      module Roda
+        # Instrumentation for Roda
+        module Instrumentation
+          def _roda_handle_main_route
+            instrument(Ext::SPAN_REQUEST) { super }
+          end
+
+          def call
+            instrument(Ext::SPAN_REQUEST) { super }
+          end
+
+          private
+
+          def instrument(span_name, &block)
+            set_distributed_tracing_context!(request.env)
+
+            Tracing.trace(span_name) do |span|
+              begin
+                request_method = request.request_method.to_s.upcase
+
+                span.service = configuration[:service_name]
+                span.span_type = Tracing::Metadata::Ext::HTTP::TYPE_INBOUND
+
+                # Using the http method as a resource, since the URL/path can trigger
+                # a possibly infinite number of resources.
+                span.resource = request_method
+
+                span.set_tag(Tracing::Metadata::Ext::HTTP::TAG_URL, request.path)
+                span.set_tag(Tracing::Metadata::Ext::HTTP::TAG_METHOD, request_method)
+
+                # Add analytics tag to the span
+                if Contrib::Analytics.enabled?(configuration[:analytics_enabled])
+                  Contrib::Analytics.set_sample_rate(span, configuration[:analytics_sample_rate])
+                end
+
+                # Measure service stats
+                Contrib::Analytics.set_measured(span)
+              ensure
+                begin
+                  response = yield
+                rescue
```

Comment:
Generally recommend classifying what kind of error is caught: `StandardError` if you don't want to trap kill signals, `Exception` if you want all signals (have be to very careful when catching `Exception`). `StandardError` is likely correct here.

To clarify, the purpose here is to add additional tags in case of an error? I think it might be better to provide an `on_error` argument to the `#trace` call, as it will handle the exception and what not for you. this way you can eliminate the `begin/rescue` in the `ensure`. Bonus is that it will handle `Exception` for you.

Reply:
I paired with Wan on this one as well, and the `on_error` would not fit well for this case because we need the local closure variable `request_method` in the error handler.

If we extract the error handling to `on_error: method(:handle_error)`, then `def handle_error` does not have access to `request_method`.
Alternatively, we could declare it inlince `on_error: ->{...}` but the method is pretty large.

I think we can keep it as in this file, as the alternatives are less legible.

Reply:
I paired with Wan on this one as well, and the `on_error` would not fit well for this case because we need the local closure variable `request_method` in the error handler.

If we extract the error handling to `on_error: method(:handle_error)`, then `def handle_error` does not have access to `request_method`.
Alternatively, we could declare it inlince `on_error: ->{...}` but the method is pretty large.

I think we can keep it as in this file, as the alternatives are less legible.